providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,631 @@
+        if (p->data != NULL) {
+            if (p->data_size < p->return_size)
+                return 0;
+            if (!ossl_ml_kem_vencode_public_key(v, p->data, p->return_size,
+                                                key->pubkey, key->prvkey))
+                return 0;
+        }
+    }
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PRIV_KEY);
+    if (p != NULL && key->prvkey != NULL) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        p->return_size = key->vinfo->prvkey_bytes;
+        if (p->data != NULL) {
+            if (p->data_size < p->return_size)
+                return 0;
+            if (!ossl_ml_kem_vencode_private_key(v, p->data, p->return_size,
+                                                 key->prvkey))
+                return 0;
+        }
+    }
+    return 1;
+}
+
+static const OSSL_PARAM *ml_kem_settable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+static int ml_kem_set_params(void *vkey, const OSSL_PARAM params[])
+{
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
+    vinfo_t v = key->vinfo;
+    const OSSL_PARAM *p;
+    const void *pubenc = NULL;
+    size_t publen = 0;
+
+    if (params == NULL)
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
+        return 1;
+
+    /* Key mutation is reported generally not allowed */
+    if (have_keys(key)) {
+        ERR_raise_data(ERR_LIB_PROV,
+                       PROV_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE,
+                       "ML-KEM keys cannot be mutated");
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
+        return 0;
+
+    if (publen != v->pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+    if (ossl_ml_kem_vparse_public_key(v, &key->pubkey, pubenc, publen,
+                                      key->ctx) == 1)
+        return 1;
+
+    OPENSSL_free(key->pubkey);
+    key->pubkey = NULL;
+    ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+    return 0;
+}
+
+static int ml_kem_gen_set_params(void *vgctx, const OSSL_PARAM params[])
+{
+    struct ml_kem_gen_ctx *gctx = vgctx;
+    const OSSL_PARAM *p;
+
+    if (gctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        OPENSSL_free(gctx->propq);
+        gctx->propq = OPENSSL_strdup(p->data);
+        if (gctx->propq == NULL)
+            return 0;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ML_KEM_SEED);
+    if (p == NULL)
```

Comment:
would pref this to be inside the if block so that new params can juist be added to the end without refactoring later.

Reply:
Done in #26172